src/range_proof.rs

```diff
@@ -0,0 +1,555 @@
+            let d1 = proof.get_d1();
+            let li = proof.get_li()?;
+            let ri = proof.get_ri()?;
+
+            if li.len() != ri.len() {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not equal to vector R length".to_string(),
+                ));
+            }
+            if 1 << li.len() != commitments.len() * bit_length {
+                return Err(ProofError::InvalidLength(
+                    "Vector L length not adequate".to_string(),
+                ));
+            }
+
+            // Helper values
+            let batch_size = commitments.len();
+            let rounds = li.len();
+
+            // Batch weight (may not be equal to a zero valued scalar)
+            let weight = Scalar::random(rng);
+
+            // Start the transcript
+            let mut transcript = Transcript::new(transcript_label.as_bytes());
+            transcript.domain_separator(b"Bulletproof+", b"Range Proof");
+            transcript.validate_and_append_point(b"H", &h_base.compress())?;
+            transcript.validate_and_append_point(b"G", &h_base.compress())?;
+            transcript.append_u64(b"N", bit_length as u64);
+            transcript.append_u64(b"M", batch_size as u64);
+            for i in 0..statements[index].generators.hi_base().len() {
+                transcript.validate_and_append_point(
+                    b"Hi",
+                    &statements[index].generators.hi_base()[i].compress(),
+                )?;
+            }
+            for i in 0..statements[index].generators.gi_base().len() {
+                transcript.validate_and_append_point(
+                    b"Gi",
+                    &statements[index].generators.gi_base()[i].compress(),
+                )?;
+            }
+            for i in 0..(statements[index].commitments.len()) {
+                transcript.validate_and_append_point(
+                    b"Ci",
+                    &statements[index].commitments[i].compress(),
+                )?;
+            }
+
+            // Reconstruct challenges
+            transcript.validate_and_append_point(b"H", &proof.a)?;
+            let y = transcript.challenge_scalar(b"y");
+            let z = transcript.challenge_scalar(b"z");
+            transcript.domain_separator(b"Bulletproof+", b"Inner Product Proof");
+            let mut challenges = vec![];
+            let mut challenges_inv = vec![];
+            for j in 0..rounds {
+                transcript.validate_and_append_point(b"L", &li[j].compress())?;
+                transcript.validate_and_append_point(b"R", &ri[j].compress())?;
+                let e = transcript.challenge_scalar(b"e");
+                challenges.push(e);
+                challenges_inv.push(e.invert());
+            }
+            transcript.validate_and_append_point(b"A1", &a1.compress())?;
+            transcript.validate_and_append_point(b"B", &b.compress())?;
+            let e = transcript.challenge_scalar(b"e");
```

Comment:
The challenges must be nonzero. This is exceptionally unlikely, but it's inexpensive to check this.